proposals/NNNN-outofproc-compiler-api-architecture.md

```diff
@@ -0,0 +1,408 @@
+<!-- {% raw %} -->
+
+# Out-of-process compiler api architecture
+
+* Proposal: [NNNN](NNNN-outofproc-compiler-api-architecture.md)
+* Author(s): [Cooper Partin](https://github.com/coopp)
+* Sponsor: [Cooper Partin](https://github.com/coopp)
+* Status: **Under Consideration**
+* Impacted Project(s): (Clang)
+
+* Issues:
+
+## Introduction
+
+An effort is underway to bring compilation support into clang for HLSL based
+shaders.  A C style api will be built to enable applications to compile
+a shader from their own processes in addition to being able to launching the
+clang process.
+
+This document is to propose a more detailed architecture for the approved 
+proposal: [0005](0005-inproc-outofproc-compiler-api-support.md).
+
+The c api [0008](0008-c-interface-compiler-library.md) for compiling a shader
+will be implemented using this out of process architecture.
+
+## Proposed solution
+
+The architecture for an out of process design will take advantage of systems
+that have multiple processors, or multiple-core processors. A process pool will
+be created that creates a separate compiler process for each available
+processor. For example, if the system has four processors, then four compiler
+processes are created.
+
+The process pool is owned by a singleton object and shared between each api
+instance created in the calling process.
+
+A set of low-level apis will be designed that will drive compilation work
+across multiple processes.
+
+> What do these apis look like and how are they called?
+
+Having a separate low-level api for this design enables other tooling clients
+to build their own high-level systems and implement their own policies.
+
+The high-level HLSL compiler c api implementation will be built on top of the
+low-level c apis and implement any required synchronous and asynchronous api
+policy behaviors. The high-level c api call will be blocked either waiting for
+a worker process to become available or an already dispatched work to be
+completed. This behavior depends on the type of call used
+(synchronous/asynchronous).
+
+Communication with the process pool occurs in-process to the application
+and the pool's monitoring threads use a named pipe IPC mechanism with a
+protocol to communicate with each monitored clang process.
+
+## Detailed design
+
+> TBD: Design the low-level apis that support this architecture and include
+code snippets of how they will be called.
+
+### Out-of-process system initialization
+
+On first creation of a low-level compiler api instance, a singleton object is
+created in-process to the caller.
+
+The singleton object manages all state and request traffic from the calling
+process. The singleton contains an api dispatching system that interfaces with
+a process pool. Each worker process in the pool is monitored by a dedicated
```

Comment:
> The singleton contains an api dispatching system that interfaces with a process pool.

This is quite different from the design I imagined.  As soon as you have a process pool you've got a whole load of extra policy involved which makes the low-level API quite high level.  I was imagining more that the low-level API is maybe responsible for working with individual clang processes, and is where the RPC protocol is implemented.

Something like: (although trying to mix components and API layering doesn't really work here)

```mermaid
flowchart TD

    subgraph HLAPI["High-level API"]
        ApiBehaviorLogic("Api Behavior Logic<br/>(implements sync/async, packs/unpacks parameters)")
        ApiCallDispatcher{"Api Dispatcher<br/>(find available process and dispatch work)"}

        subgraph CompilerProcessPool["Compiler Process Pool"]
        subgraph MonitorThread1["Monitor Thread (1)"]
            subgraph JsonRPC ["Low-Level API"]
            end
        end
        subgraph MonitorThread2["Monitor Thread (2)"]
            subgraph JsonRPC2 ["Low-level API"]
            end
        end
        end

    end



    subgraph Process2[" "]
        ClangProcess2["clang.exe (2)"]
        db2[("compiled<br/>shader")]
    end

    subgraph Process1[" "]
        ClangProcess1["clang.exe (1)"]
        db[("compiled<br/>shader")]
    end

    

    style Process1 stroke-dasharray:10,10
    style Process2 stroke-dasharray:10,10


    CompilerApi[Compiler Api] -->|"Compile/CompileAsync( )"| ApiBehaviorLogic

    ApiBehaviorLogic <--> |"dispatch and monitor call"| ApiCallDispatcher

    ApiCallDispatcher -->JsonRPC<--> | json-rpc<br/>protocol | ClangProcess1 --> db

    ApiCallDispatcher --> JsonRPC2<-->| json-rpc<br/>protocol | ClangProcess2 --> db2
```